### PR TITLE
add napi file, modify makefile, for CreateKey.

### PR DESCRIPTION
**Pull request by Liang1XZhao**
add napi file, implementation CreateKey function

### CODE REVIEW

File: core/Napi/ehsm_napi.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        //char [3]
+        output[3] = (char)(input[2] & 0x3f);
+        pDest[index_dest++] = BASE64_ENCHARS[output[3]];
+    }
+    //null-terminator
+    pDest[index_dest] = '\n';
+    return index_dest;
+}
+
+extern "C" char* CreateKey_napi(int intMechanism, int intOrigin)
```
Comment (syan10) [inline core/Napi/ehsm_napi.cpp:95]: why the return code is "char *", is it must?

File: core/Napi/ehsm_napi.cpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+{
+    EH_RV rv = EHR_FUNCTION_FAILED;
+    EH_KEY_BLOB key_blob;
+    char* rest;
+    int i = 0;
+
+    key_blob.pKeyData = NULL;
+    key_blob.ulKeyLen = 0;
+    
+    rv = Initialize();
```
Comment (syan10) [inline core/Napi/ehsm_napi.cpp:105]: not call the finalize()?

File: core/App/ehsm_provider.cpp [hunk 3]
```diff
@@ -108,22 +108,28 @@ EH_RV CreateKey(EH_MECHANISM_TYPE ulKeySpec, EH_KEY_ORIGIN eOrigin,
...
+    printf("==========CreateKey IN==========\n");
+    printf("ulKeySpec : %lu\n", ulKeySpec);
+    printf("eOrigin : %d\n", eOrigin);
+    printf("pKeyBlob : %p\n", pKeyBlob);
 
     if (eOrigin != EHO_INTERNAL_KEY || pKeyBlob == NULL) {
         return EHR_ARGUMENTS_BAD;
     }
-
+    printf("==========CreateKey 1==========\n");
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:119]: remove the unused logs

File: core/App/ehsm_provider.cpp [hunk 4]
```diff
@@ -108,22 +108,28 @@ EH_RV CreateKey(EH_MECHANISM_TYPE ulKeySpec, EH_KEY_ORIGIN eOrigin,
...
     pKeyBlob->ulKeyType = ulKeySpec;
 
     switch (ulKeySpec) {
         case EHM_AES_GCM_128:
             if (pKeyBlob->pKeyData == NULL) {
+                printf("==========CreateKey 2.1==========\n");
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:125]: same here

File: core/Napi/ehsm_napi.cpp.work [hunk 5]
```diff
@@ -0,0 +1,287 @@
+/*
```
Comment (syan10) [inline core/Napi/ehsm_napi.cpp.work:1]: this file for what?

File: core/Makefile [hunk 6]
```diff
@@ -34,6 +34,7 @@ include ../buildenv.mk
 App_Name := ehsm-core
```
Comment (syan10) [inline core/Makefile:34]: if you already split it to the libaray, please remove the original App target.

**Review by syan10 — CHANGES_REQUESTED** [review]

**Review by syan10 — CHANGES_REQUESTED** [review]
And, you should split this PR into serveral commits.

Only do one thing in one comimt.

File: core/App/ehsm_provider.cpp [hunk 7]
```diff
@@ -63,7 +63,7 @@ EH_RV Initialize()
     }
 
     // create ECDH session using initiator enclave, it would create ECDH session with responder enclave running in another process
-    status = enclave_la_create_session(enclaveHelpers.getSgxEnclaveId(), &ret_status);
+    /*status = enclave_la_create_session(enclaveHelpers.getSgxEnclaveId(), &ret_status);
```
Comment (yang8621) [inline core/App/ehsm_provider.cpp:66]: keep these codes in the patch. It's for the local test only.

File: core/Napi/ehsm_napi.cpp [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            printf("create key 1 failed 0x%lx\n", rv);
+            return "create key 1 failed\n";
+        }
+    } else {
+        printf("create key 2 failed 0x%lx\n", rv);
+        return "create key 2 failed\n";
+    }
+    printf("create key done\n");
+    
+    rest = (char *)malloc(128 * sizeof(uint8_t));
```
Comment (yang8621) [inline core/Napi/ehsm_napi.cpp:131]: How to know 128 is big enough? 
Is it possible to use C++ STL and define the base64encoding as:
std::string EncodeBase64(const unsigned char* input_data, int input_num)?

Comment (yang8621) [reply core/Napi/ehsm_napi.cpp:131]: FYI:
std::string base64_encoding(const unsigned char* input_data, int input_num) {
const char encode_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
unsigned char batch_input[4] = { 0 };
std::string encode_str;
int length = 0;
for (int i = 0; i < (int)(input_num / 3); i++) {
batch_input[1] = *input_data++;
batch_input[2] = *input_data++;
batch_input[3] = *input_data++;
encode_str += encode_table[batch_input[1] >> 2];
encode_str += encode_table[((batch_input[1] << 4) | (batch_input[2] >> 4)) & 0x3F];
encode_str += encode_table[((batch_input[2] << 2) | (batch_input[3] >> 6)) & 0x3F];
encode_str += encode_table[batch_input[3] & 0x3F];



length +=4;
if (length == 76) {
encode_str += "\r\n";
length = 0;
}
}



int mod = input_num % 3;
if (mod == 2) {
batch_input[1] = *input_data++;
batch_input[2] = *input_data++;
encode_str += encode_table[(batch_input[1] & 0xFC) >> 2];
encode_str += encode_table[((batch_input[1] & 0x03) << 4) | ((batch_input[2] & 0xF0) >> 4)];
encode_str += encode_table[((batch_input[2] & 0x0F) << 2)];
encode_str += "=";
} else {
batch_input[1] = *input_data++;
encode_str += encode_table[(batch_input[1] & 0xFC) >> 2];
encode_str += encode_table[((batch_input[1] & 0x03) << 4)];
encode_str += "==";
}



return encode_str;
}

File: docker/Dockerfile [hunk 9]
```diff
@@ -61,8 +61,26 @@ COPY bin/ehsm-core/libenclave-ehsm-core.signed.so /home/ehsm/ehsm-core/libenclav
...
+    && mv node-v16.13.0-linux-x64/ /usr/local/nodejs \
+    && ln -s /usr/local/nodejs/bin/node /usr/local/bin \
+    && ln -s /usr/local/nodejs/bin/npm /usr/local/bin
+    # && npm config set registry https://registry.npm.taobao.org/
+
+COPY /ehsm-service /home/ehsm/ehsm-service
+
+RUN npm install
+
+EXPOSE 3000
```
Comment (syan10) [inline docker/Dockerfile:79]: what's this for?

File: docker/Dockerfile [hunk 10]
```diff
@@ -61,8 +61,26 @@ COPY bin/ehsm-core/libenclave-ehsm-core.signed.so /home/ehsm/ehsm-core/libenclav
 #    && cd ehsm \
 #    && make
 
+# node-service
+RUN mkdir -p /home/ehsm/ehsm-service
+WORKDIR /home/ehsm/ehsm-service
+
+RUN wget https://nodejs.org/dist/v16.13.0/node-v16.13.0-linux-x64.tar.xz \
+    && tar xf  node-v16.13.0-linux-x64.tar.xz \
+    && mv node-v16.13.0-linux-x64/ /usr/local/nodejs \
```
Comment (syan10) [inline docker/Dockerfile:70]: why not use apt-get install?  is it a prebuild nodejs?

File: ehsm-service/constans.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
+const pramas = {
```
Comment (syan10) [inline ehsm-service/constans.js:1]: change the filename "constans.js" to make more meaningful

File: ehsm-service/index.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
+const express = require('express')
```
Comment (syan10) [inline ehsm-service/index.js:1]: same to this file name..

File: ehsm-service/index.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    "EHO_INTERNAL_KEY": 0,
+    "EHO_EXTERNAL_KEY": 1
+  };
+  const intMechanism = mechanismType[mechanism];
+  const intOrigin = originType[origin];
+
+  switch (req.query.Action) {
+    case apis.CreateKey:      
+      let CreateKey_napi = ffi.Library('./libehsmnapi',{'CreateKey_napi':['string',['int','int']]});      
+      const CreateKey_r = CreateKey_napi.CreateKey_napi(intMechanism, intOrigin);
```
Comment (syan10) [inline ehsm-service/index.js:109]: so is it must to return a string for the function 'CreateKey_napi", can we put the result_string as a output parameter?

**Comment by syan10** [comment]
please create a new PR for the change, and i will abandon it now

**Pull request closed without merging**